src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -475,6 +496,54 @@ entails SolverOptions{..} constraint context hints =
       pure (arg1, arg2, srcTypeLevelString (mkString $ h' <> t'))
     consSymbol _ _ _ = Nothing
 
+    solveIsReflectable :: [SourceType] -> Maybe [TypeClassDict]
+    solveIsReflectable [typeLevel, _] = do
+      (ref, typ) <- case typeLevel of
+        TypeLevelInt _ i -> pure (ReflectableInt i, tyInt)
+        TypeLevelString _ s -> pure (ReflectableString s, tyString)
+        _ -> Nothing
+      pure [TypeClassDictionaryInScope Nothing 0 (IsReflectableInstance ref) [] C.IsReflectable [] [] [typeLevel, typ] Nothing Nothing]
+    solveIsReflectable _ = Nothing
+
+    solveIntAdd :: [SourceType] -> Maybe [TypeClassDict]
+    solveIntAdd [arg0, arg1, arg2] = do
+      (arg0', arg1', arg2') <- addInts arg0 arg1 arg2
+      let args' = [arg0', arg1', arg2']
+      pure [TypeClassDictionaryInScope Nothing 0 EmptyClassInstance [] C.IntAdd [] [] args' Nothing Nothing]
+    solveIntAdd _ = Nothing
+
+    addInts :: SourceType -> SourceType -> SourceType -> Maybe (SourceType, SourceType, SourceType)
+    -- | l r -> o, l + r = o
+    addInts arg0@(TypeLevelInt _ l) arg1@(TypeLevelInt _ r) _ = pure (arg0, arg1, srcTypeLevelInt (l + r))
+    -- | l o -> r, o - l = r
+    addInts arg0@(TypeLevelInt _ l) _ arg2@(TypeLevelInt _ o) = pure (arg0, srcTypeLevelInt (o - l), arg2)
+    -- | r o -> l, o - r = l
+    addInts _ arg1@(TypeLevelInt _ r) arg2@(TypeLevelInt _ o) = pure (srcTypeLevelInt (o - r), arg1, arg2)
+    addInts _ _ _                                             = Nothing
+
+    solveIntCompare :: InstanceContext -> [SourceType] -> Maybe [TypeClassDict]
+    solveIntCompare _ [arg0@(TypeLevelInt _ a), arg1@(TypeLevelInt _ b), _] =
+      let ordering = case compare a b of
+            EQ -> C.orderingEQ
+            LT -> C.orderingLT
+            GT -> C.orderingGT
+          args' = [arg0, arg1, srcTypeConstructor ordering]
+      in pure [TypeClassDictionaryInScope Nothing 0 EmptyClassInstance [] C.IntCompare [] [] args' Nothing Nothing]
+    solveIntCompare ctx [a, b, _] = do
+      let compareDictsInScope = findDicts ctx C.IntCompare Nothing
+          givens = flip mapMaybe compareDictsInScope $ \case
+            dict | [a', b', c'] <- tcdInstanceTypes dict -> mkRelation a' b' c'
+                 | otherwise -> Nothing
+      c' <- solveRelation givens a b
```

Comment:
There's one more enhancement that I think would be important to cover:

```purs
litTrans :: forall a. Compare a 1 GT => Proxy ( left :: a, right :: 0 )
litTrans = assertGreater
```

Because the compiler can solve arithmetic facts like `Compare 1 0 GT`, and the compiler can make transitive connections between `Compare` dictionaries in the context, I think for consistency we should make sure the compiler can also make transitive connections between context dictionaries and arithmetic facts.

One way I can imagine doing this is adding to the list of givens here the full set of `Unequal`s that can be inferred from the distinct `TypeLevelInt`s that are used in the `compareDictsInScope`, `a`, and `b`. It's brutish, but should suffice—I have a hard time imagining having so many known integer types floating around in `Compare` dicts at once that the extra O(n^2) relations cause any noticeable problems—at least, not before the core `solveRelation` implementation does. But if you can devise a more elegant approach, please be my guest.

Reply:
> Could you elaborate on this approach?

Suppose, as in your second example, you have the context `a < 10, 0 < a` and you are in the process of solving `Compare a 20 o`. My concept is:
* Collect all the integer literals under discussion and sort: `[0, 10, 20]`
* Iterating over the `tails` of that list, add nominal-style constraints to the context: `0 < 10, 0 < 20, 10 < 20`
* Solve as usual

I'm not quite sure I fully understand *your* approach, but I'm guessing it'd require more code and it's not yet clear to me that it's more efficient.

Reply:
What I'm thinking of is basically something like:
```hs
module Main where

import Prelude

import Data.Foldable (all)
import Effect (Effect)
import Effect.Console (logShow)

data Relationship a
  = LessThan a
  | GreaterThan a

type Context a = Array (Relationship a)

solveFacts :: forall a. Ord a => Context a -> Relationship a -> Boolean
solveFacts context relationship = all (relationship `adheresTo` _) context
  where
  adheresTo :: Relationship a -> Relationship a -> Boolean
  adheresTo (LessThan a) (LessThan b) = b > a
  adheresTo (GreaterThan a) (GreaterThan b) = a > b
  adheresTo (LessThan a) (GreaterThan b) = a > b
  adheresTo (GreaterThan a) (LessThan b) = b > a

main :: Effect Unit
main = do
  logShow $ solveFacts [GreaterThan 0, LessThan 10] (GreaterThan 5)
```
The compiler would collect arithmetic facts to build the context for a specific variable, then verify the fact that's being checked. Though this approach would likely need correction.